### PR TITLE
aws.ebs-snapshot - copy-volume-tags action

### PR DESCRIPTION
**Pull request by thisisshi**
Resolves: https://github.com/cloud-custodian/cloud-custodian/issues/2497

### CODE REVIEW

File: c7n/resources/ebs.py [hunk 1]
```diff
@@ -218,6 +218,79 @@ def process(self, snapshots, event=None):
...
+        volume_ids = [r['VolumeId'] for r in resources]
+
+        while volume_ids:
+            try:
+                vols = client.describe_volumes(VolumeIds=volume_ids)
+                break
+            except ClientError as e:
+                if e.response['Error']['Code'] == 'InvalidVolume.NotFound':
+                    if not self.data.get('skip-missing-volumes', True):
+                        raise PolicyExecutionError("Snapshots not found: %s" % e)
```
Comment (thisisshi) [inline c7n/resources/ebs.py]: Needs a more descriptive/accurate message here oops

File: c7n/resources/ebs.py [hunk 2]
```diff
@@ -218,6 +218,89 @@ def process(self, snapshots, event=None):
...
+
+    def process(self, resources):
+        client = local_session(self.manager.session_factory).client('ec2')
+        tag_keys = self.data.get('tags', [])
+        volume_ids = [r['VolumeId'] for r in resources]
+
+        volume_tag_map = self.get_volume_tag_map(volume_ids)
+
+        for r in resources:
+            if r['VolumeId'] in volume_tag_map:
```
Comment (kapilt) [inline c7n/resources/ebs.py]: can dedent this block by handling the not found case with a continue at the top of the loop
ie 
```python
if r['VolumeId'] not in volume_tag_map:
    continue
```

File: c7n/resources/ebs.py [hunk 3]
```diff
@@ -218,6 +218,89 @@ def process(self, snapshots, event=None):
...
+                vol_tags = volume_tag_map[r['VolumeId']]
+                if vol_tags:
+                    add_tags = [{'Key': t['Key'], 'Value': t['Value']} for t in vol_tags if t['Key'] in tag_keys] # noqa
+                else:
+                    continue
+                vol_tags.extend(add_tags)
```
Comment (kapilt) [inline c7n/resources/ebs.py]: need to sanity check that there's actually something changed before doing the api call, else we'll be continually doing api calls without any actual work to perform.

File: c7n/resources/ebs.py [hunk 4]
```diff
@@ -218,6 +218,89 @@ def process(self, snapshots, event=None):
...
+                    'Unable to find volume: %s for snapshot: %s' % (r['VolumeId'], r['SnapshotId']))
+
+    def get_volume_tag_map(self, volume_ids):
+        """
+        Returns a mapping of {VolumeIds: [Tags]}
+        """
+        client = local_session(self.manager.session_factory).client('ec2')
+        while volume_ids:
+            try:
+                vols = client.describe_volumes(VolumeIds=volume_ids)
```
Comment (kapilt) [inline c7n/resources/ebs.py]: use the volume resource manager, that will respect caches, sources etc.

File: c7n/resources/ebs.py [hunk 5]
```diff
@@ -218,6 +218,77 @@ def process(self, snapshots, event=None):
...
+                self.log.warning(
+                    'Unable to find volume: %s for snapshot: %s' % (r['VolumeId'], r['SnapshotId']))
+                continue
+            vol_tags = volume_tag_map[r['VolumeId']]
+            if vol_tags:
+                add_tags = [{'Key': t['Key'], 'Value': t['Value']} for t in vol_tags if t['Key'] in tag_keys] # noqa
+            else:
+                continue
+            if add_tags:
+                vol_tags.extend(add_tags)
```
Comment (kapilt) [inline c7n/resources/ebs.py]: The volume tags should be intersected with the snapshot tags to find a delta of tag values to change, else as is, everytime this runs its going to set the same tag values on a snapshot that already has those values.

File: c7n/resources/ebs.py [hunk 6]
```diff
@@ -218,6 +218,77 @@ def process(self, snapshots, event=None):
...
+                    - Owner
+    """
+
+    permissions = ('ec2:DescribeVolumes', 'ec2:CreateTags')
+    schema = type_schema(
+        'copy-volume-tags',
+        required=['tags'],
+        **{
+            'skip-missing-volumes': {'type': 'boolean'},
+            'tags': {'type': 'array', 'items': {'type': 'string'}}
```
Comment (kapilt) [inline c7n/resources/ebs.py]: thoughts on a "*", "all" value?

**Comment by thisisshi** [comment]
https://github.com/cloud-custodian/cloud-custodian/compare/master...thisisshi:generic-copy-related-resource-tag

wip implementation of the above

could also see an action "copy-tag-to-related-resource" as well or maybe within the same action? some resource relations only exist in one direction, e.g. ec2 -> iam instance profile

**Comment by thisisshi** [comment]
@kapilt yea key_id is mainly to prevent the need to re-implement the related tag filter over and over again, the action itself would just be automatically added to any resource that has the tag action.

re: skip_missing-- yes would default to true, mostly just there for documentation
re: bidirectionality-- gotcha, was trying to think of a more realistic use case but couldn't

**Comment by thisisshi** [comment]
Super-ceded by: https://github.com/cloud-custodian/cloud-custodian/pull/3489

**Pull request closed without merging**